docs/job_options.md
# Tweak batch jobs with `job_options`

The Geotrellis openEO back-end supports passing additional options at job creation time to tweak batch jobs:

```python
connection.create_job(process_graph, additional={
    "driver-memory": "8G",
    "driver-memoryOverhead": "2G"
})
```

```python
data_cube.create_job(job_options={
    "sentinel-hub": {
        "client-alias": "vito",
        "input": "sync"
    }
})
```

### Log level

To reduce:
* the amount of irrelevant logs for users to wade through,
* the amount of storage for these logs on disk,

the openEO API spec provides a `log_level` to set at job creation time,
for example with `create_job()` or `execute_batch()`:

```python
cube.create_job(
    log_level="warning",
)
```

(Note that, originally, this log level could also be set through a "logging-threshold" job option,
but that usage pattern is deprecated in favor of the standardized way mentioned above)

This log level effectively prevents log entries with a lower level (in this case: "debug" and "info") from being written to the logs.

It accepts the log levels as defined in the OpenEO API: "debug", "info", "warning" and "error" and defaults to "info".

Note: logs originating from ```openeo.udf.debug.inspect``` are unaffected by this threshold; regardless of their log
level, be it "debug", "info", "warning" or "error", they will always be written to the logs.

### Allow empty data cubes

Setting the "allow_empty_cubes" flag (defaults to: `false`) enables working with empty data cubes
where this otherwise would raise an error. It is applied to all collections within this job.

```python
job_options={
    "allow_empty_cubes": True
}
```


docs/configuration.md

# Configuration

## Background

The openEO GeoPySpark backend grew organically from a single instance proof of concept
to a component that is used in different deployments.
As a result, the (need for) configuration options also grew organically,
and the current implementation uses a mix of configuration approaches of varying flexibility:

- Just **hardcoded values** directly in the code that needs it.
  Obviously, this is all but flexible and needs to be avoided.
- Getting a configuration parameter from **environment variables directly**
  in the code that needs it (e.g. with `os.environ.get()` in Python).
  This allows for some basic flexibility, but
  the configuration option is hidden deep in the code and cumbersome to maintain
  (duplication, management of fallback values, ...)
- The **`ConfigParams` class** (in `openeogeotrellis/configparams.py`) was the first approach
  to have a **centralized** place for configuration parameters
  and improve the maintenance and overview challenges.

`ConfigParams` is mostly still based on reading and parsing environment variables.
This is fine for basic things like the URL of some service or database,
but it doesn't scale very well:

- the environment variables tend to be sprinkled all over the place:
  shell scripts, Python startup scripts, Jenkinsfiles, Dockerfiles, etc.,
  possibly across different repositories.
  It's hard to figure out where to find, add or update a particular setting.
- environment variables only work well for string parameters,
  but some configuration requires more complex constructs (lists, mappings, nested constructs)
  and trying to properly encode these in a single string as env variable is not always trivial
  (e.g. it must survive possible string manipulation in shell scripts along the way).


## Config files

To address the limitations of doing configuration through environment variables,
(yet) another configuration system was introduced
(under [#285](https://github.com/Open-EO/openeo-geopyspark-driver/issues/285))
based on configuration files.
**Python** was chosen as configuration file format (instead of JSON, YAML, TOML, ...)
for maximum flexibility and easy integration with the openeo-geopyspark-driver and related code
(e.g. type hints and code navigation features).
Configuration parameters can be defined directly as Python values,
not only the usual suspects like lists, dicts, tuples, sets, etc.
but also custom objects or constructs.
It also allows to derive parameters dynamically,
for example based on environment variables or other contextual information.

Note that this file based configuration system is introduced gradually
and the original `ConfigParams` is largely kept untouched.


### Usage

#### Setup

- A config file with this system is expected to define a `config` variable at top-level,
  and it must be an instance of the class `openeogeotrellis.config.GpsBackendConfig`.
  A minimal config file looks like this:
  ```python
  from openeogeotrellis.config import GpsBackendConfig
  config = GpsBackendConfig()
  ```
- When starting the openEO GeoPyspark backend,
  the environment variable `OPENEO_BACKEND_CONFIG` must be set
  contain the path to this config file.

Also look at the [default config file](https://github.com/Open-EO/openeo-geopyspark-driver/blob/master/openeogeotrellis/config/default.py)
which is loaded when `OPENEO_BACKEND_CONFIG` is not set.


#### Setting values

The `GpsBackendConfig` class is defined through
the popular [`attrs`](https://www.attrs.org/en/stable/) package
to compactly define attributes with type hints and default values.

For example:

```python
from openeo_driver.users.oidc import OidcProvider

@attrs.define
class GpsBackendConfig:
    oidc_providers: List[OidcProvider] = attrs.Factory(list)
```

Config parameter `oidc_providers` must be a list of `OidcProvider` instances,
and will be an empty list by default.
A custom config file that sets this parameter might look like this:

```python
from openeo_driver.users.oidc import OidcProvider
from openeogeotrellis.config import GpsBackendConfig

config = GpsBackendConfig(
    oidc_providers=[
      OidcProvider(id="OIDC", issuer="https://oidc.test", title="test OIDC"),
    ],
)
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

This project relies on continuous integration for new features. So we do not yet have explicitly versioned
releases. Releases are simply built continuously, automatically tested, deployed to a development environment and then to production.

Note that the openEO API provides a way to support stable and unstable versions in the same implementation:
https://openeo.org/documentation/1.0/developers/api/reference.html#operation/connect

If needed, feature flags are used to allow testing unstable features in development/production,
without compromising stable operations.

<!-- start-of-changelog -->

## In progress: 0.65.0

- `save_result`: write GeoTiff assets with valid tile size; override with `tile_size` format option. ([#1083](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1083))
- `sar_backscatter` support: automatically use backend-specific `coefficient` default even if there is no explicit `sar_backscatter` in the process graph ([Open-EO/openeo-python-driver#376](https://github.com/Open-EO/openeo-python-driver/issues/376))
- Add `capabilities_extras` config to easily extend capabilities document ([Open-EO/openeo-python-driver#384](https://github.com/Open-EO/openeo-python-driver/issues/384))
- Add `CalrissianS3Result.generate_presigned_url()` ([#937](https://github.com/Open-EO/openeo-geopyspark-driver/issues/937))
- Restore COG layout for `GTiff` output format ([Open-EO/openeo-geotrellis-extensions#393](https://github.com/Open-EO/openeo-geotrellis-extensions/issues/393))
- Add `CalrissianS3Result.download()` ([#937](https://github.com/Open-EO/openeo-geopyspark-driver/issues/937))
- Support additional Sentinel 3 collections ([eu-cdse/openeo-cdse-infra#380](https://github.com/eu-cdse/openeo-cdse-infra/issues/380))
- Add `CalrissianS3Result.generate_public_url()` ([#937](https://github.com/Open-EO/openeo-geopyspark-driver/issues/937))
- Replace `ConfigParams.layer_catalog_metadata_files` with `GpsBackendConfig.layer_catalog_files` ([#285](https://github.com/Open-EO/openeo-geopyspark-driver/issues/285), [#1084](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1084))
- `StacApiWorkspace`: log body of error response from STAC API for better root cause analysis ([#1116](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1116))
- `export_workspace`: add `"derived_from"` links to STAC Collection ([#1050](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1050))
- Calrissian integration: avoid unnecessary pulls of `alpine` image ([#1132]https://github.com/Open-EO/openeo-geopyspark-driver/issues/1132)
- Calrissian integration: refactor config to a `CalrissianConfig` sub-config ([#1009](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1009))
- `save_result`: allow non-string values in `GTiff` `file_metadata` ([#1142](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1142))
- Add `udp_registry_zookeeper_client_reuse` config for `KazooClient` reuse in `ZooKeeperUserDefinedProcessRepository` ([#1037](https://github.com/Open-EO/openeo-geopyspark-driver/pull/1037))
- `GpsBackendConfig`: be more forgiving about unknown config keys to better support use cases that involve backward/forward incompatible configurations ([Open-EO/openeo-python-driver#322](https://github.com/Open-EO/openeo-python-driver/issues/322))


## 0.64.1

- `load_custom_processes`: allow specifying path directly


## 0.63.0

- Add `bbox` and `geometry` to exported STAC items pertaining to GeoTiff assets of a spatial data cube ([eu-cdse/openeo-cdse-infra#418](https://github.com/eu-cdse/openeo-cdse-infra/issues/418))
- Support UDF dependency extraction from remote process definitions (URL based UDPs) ([#1063](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1063))
- `StacApiWorkspace`: improve STAC requests resilience ([#1073](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1073))
- `StacApiWorkspace`: reject unsupported characters in `merge` ([eu-cdse/openeo-cdse-infra#415](https://github.com/eu-cdse/openeo-cdse-infra/issues/415))
- `load_stac`: fix delay in driver due to expensive item/asset processing ([#1081](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1081))


## 0.62.0

- `load_stac`: improve STAC requests resilience ([#818](https://github.com/Open-EO/openeo-geopyspark-driver/issues/818))
- Extract demo CWL content to package resources ([#1057](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1057))
- Disable EJR health check from batch job context ([#1066](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1066))
- `load_stac`: support empty data cubes ([#1049](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1049))
- Remove proof-of-concept CWL processes (now in openeo-geotrellis-kubernetes) from generic openeo-geopyspark-driver ([#1057](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1057)/[#1038](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1038))


## 0.61.0

- `python-memory`: make job option available on YARN, add a default config
- `load_stac`: optimize resolution with regard to requested bands (experimental) ([#1043](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1043))
- `load_stac`: apply offset (experimental) ([#1051](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1051))
- Deprecate non-standard "logging-threshold" job option in favor of standardized "log_level" job creation parameter ([#909](https://github.com/Open-EO/openeo-geopyspark-driver/issues/909))
- Fail fast on UDF dependency installation failure ([#1048](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1048))
- `load_stac`: avoid OOM on global low-res assets ([#1055](https://github.com/Open-EO/openeo-geopyspark-driver/issues/1055))

## 0.60.1


openeogeotrellis/config/config.py
from __future__ import annotations

import abc
import os
from pathlib import Path
from typing import List, Optional, Union

import attrs
from openeo_driver.config import OpenEoBackendConfig, from_env_as_list
from openeo_driver.users import User
from openeo_driver.users.oidc import OidcProvider
from openeo_driver.util.auth import ClientCredentials
from openeo_driver.utils import smart_bool

from openeogeotrellis import get_backend_version
from openeogeotrellis.config.constants import UDF_DEPENDENCIES_INSTALL_MODE
from openeogeotrellis.config.integrations.calrissian_config import CalrissianConfig
from openeogeotrellis.deploy import (
    build_gps_backend_deploy_metadata,
    find_geotrellis_jars,
)


def _default_capabilities_deploy_metadata() -> dict:
    metadata = build_gps_backend_deploy_metadata(
        packages=[
            "openeo",
            "openeo_driver",
            "openeo-geopyspark",
            "geopyspark",
        ],
        jar_paths=find_geotrellis_jars(),
    )
    return metadata


class EtlApiConfig(metaclass=abc.ABCMeta):
    """
    Interface for configuration of ETL API access (possibly dynamic based on user, ...).

    The basic design idea of this interface is to use the ETL API root URL as full identifier of an ETL API.
    Being a simple string, instead of a complex object (e.g. an `EtlApi` instance), it does not raise
    serialization challenges when there is need to pass it from the web app context to the batch job tracker context.

    The ETL API selection strategy is to be implemented in `get_root_url()`,
    which can use the provided user object or job options to pick the appropriate ETL API root URL.
    Note that there is no guarantee that either of these inputs is available.
    Typically, the job options will be provided in a job tracker context
    and the user object will be provided in a synchronous processing request context.

    Additional dependencies to (re)construct an operational `EtlApi` instance can be obtained with dedicated methods
    using the ETL API root URL identifier, e.g. client credentials with `get_client_credentials(root_url)`.
    """

    @abc.abstractmethod
    def get_root_url(self, *, user: Optional[User] = None, job_options: Optional[dict] = None) -> str:
        """Get root URL of the ETL API"""
        ...

    def get_client_credentials(self, root_url: str) -> Union[ClientCredentials, None]:
        """Get client credentials corresponding to root URL."""
        return None


@attrs.frozen(kw_only=True)
class GpsBackendConfig(OpenEoBackendConfig):
    """
    Configuration for GeoPySpark backend.

    Meant to gradually replace ConfigParams (Open-EO/openeo-geopyspark-driver#285)
    """

    # identifier for this config
    id: Optional[str] = None

    capabilities_backend_version: str = get_backend_version()
    capabilities_deploy_metadata: dict = attrs.Factory(_default_capabilities_deploy_metadata)
    capabilities_extras: Optional[dict] = None
    processing_software = f"openeo-geopyspark-driver-{get_backend_version()}"

    oidc_providers: List[OidcProvider] = attrs.Factory(list)

    layer_catalog_files: List[str] = os.environ.get("OPENEO_CATALOG_FILES", "layercatalog.json").split(",")

    # Temporary feature flag for preventing to run UDFs in driver process (https://github.com/Open-EO/openeo-geopyspark-driver/issues/404)
    # TODO: remove this temporary feature flag
    allow_run_udf_in_driver: bool = False

    # TODO: avoid KUBE env var and just default to False in general
    setup_kerberos_auth: bool = not smart_bool(os.environ.get("KUBE", False))

    # TODO: possible to disable enrichment by default?
    opensearch_enrich: bool = True
    # TODO: eliminate hardcoded VITO/Terrascope resources
    default_opensearch_endpoint: str = "https://services.terrascope.be/catalogue"

    # TODO: eliminate hardcoded VITO-specific defaults?
    logging_es_hosts: List[str] = os.environ.get("LOGGING_ES_HOSTS", "https://es-infra.vgt.vito.be").split(",")
    # TODO: this index pattern is specifically used for fetching batch job logs, which is not obvious from the name (and env var)
    logging_es_index_pattern: str = os.environ.get("LOGGING_ES_INDEX_PATTERN", "openeo-*-index-1m*")

    # TODO eliminate hardcoded VITO reference
    vault_addr: Optional[str] = os.environ.get("VAULT_ADDR", "https://vault.vgt.vito.be")

    zookeeper_hosts: List[str] = attrs.Factory(
        from_env_as_list(
            "ZOOKEEPERNODES",
            # TODO: eliminate hardcoded default once config is set where necessary
            default="epod-master1.vgt.vito.be:2181,epod-master2.vgt.vito.be:2181,epod-master3.vgt.vito.be:2181",
        )
    )
    zookeeper_root_path: str = attrs.field(
        default="/openeo", validator=attrs.validators.matches_re("^/.+"), converter=lambda s: s.rstrip("/")
    )

    # TODO #236/#498/#632 long term goal is to fully disable ZK job registry, but for now it's configurable.
    use_zk_job_registry: bool = True
    zk_job_registry_max_specification_size: Optional[int] = None

    udp_registry_zookeeper_client_reuse: bool = False

    ejr_api: Optional[str] = os.environ.get("OPENEO_EJR_API")
    ejr_backend_id: str = "unknown"
    ejr_credentials_vault_path: Optional[str] = os.environ.get("OPENEO_EJR_CREDENTIALS_VAULT_PATH")

    # TODO: eliminate hardcoded Terrascope references
    # TODO #531 eliminate this config favor of etl_api_config strategy below
    etl_api: Optional[str] = os.environ.get("OPENEO_ETL_API", "https://etl.terrascope.be")
    etl_source_id: str = "TerraScope/MEP"
    use_etl_api_on_sync_processing: bool = False
    etl_dynamic_api_flag: Optional[str] = None  # TODO #531 eliminate this temporary feature flag? Unused now

    # TODO #531 this config is meant to replace `etl_api` from above
    etl_api_config: Optional[EtlApiConfig] = None

    prometheus_api: Optional[str] = os.environ.get("OPENEO_PROMETHEUS_API")

    max_executor_or_driver_memory: str = "64G"  # Executors and drivers have the same amount of memory

    # TODO #734: these "usage" defaults are a hack in absence of more accurate credit attribution
    #   for synchronous processing. Can they be eliminated?
    default_usage_cpu_seconds: float = 1.5 * 3600
    default_usage_byte_seconds: float = 3 * 1024 * 1024 * 1024 * 3600
    report_usage_sentinelhub_pus: bool = True
    batch_job_base_fee_credits: Optional[float] = None

    default_soft_errors: float = 0.1

    s1backscatter_elev_geoid: Optional[str] = os.environ.get("OPENEO_S1BACKSCATTER_ELEV_GEOID")

    # TODO: generically named config for a single bucket: not future/feature-proof (support for multiple buckets)
    s3_bucket_name: str = os.environ.get("SWIFT_BUCKET", "OpenEO-data")

    fuse_mount_batchjob_s3_bucket: bool = smart_bool(os.environ.get("FUSE_MOUNT_BATCHJOB_S3_BUCKET", False))
    fuse_mount_batchjob_s3_mounter: str = os.environ.get("FUSE_MOUNT_BATCHJOB_S3_MOUNTER", "s3fs")
    fuse_mount_batchjob_s3_mount_options: str = os.environ.get("FUSE_MOUNT_BATCHJOB_S3_MOUNT_OPTIONS", "-o uid=18585 -o gid=18585 -o compat_dir")
    fuse_mount_batchjob_s3_storage_class: str = os.environ.get("FUSE_MOUNT_BATCHJOB_S3_STORAGE_CLASS", "csi-s3")

    batch_scheduler: str = "default-scheduler"
    yunikorn_queue: str = os.environ.get("YUNIKORN_QUEUE", "root.default")
    yunikorn_scheduling_timeout: str = os.environ.get("YUNIKORN_SCHEDULING_TIMEOUT", "10800")
    yunikorn_user_specific_queues: bool = smart_bool(os.environ.get("YUNIKORN_USER_SPECIFIC_QUEUES", False))

    """
    Reading strategy for load_collection and load_stac processes:
    load_by_target_partition: first apply a partitioner, then simply read data for each partition
    This avoids moving data around after reading, allowing the reading step to be followed immediately by next processes.
    load_per_product: group by source product filename, then perform reads, then regroup data according to optimal partitioner for subsequent processes.
    This strategy is faster when there is a high overhead/latency of opening products, for instance observed when reading jpeg2000 on S3. It does require
    the data to be moved around in the cluster for subsequent processing.

    The default can be overridden by feature_flags.
    """
    default_reading_strategy: str = "load_by_target_partition"

    """
    Controls the default number of threads for the batch job executors Java Virtual Machine.
    These threads are used for additional parallelism, for instance when reading data, but may cause threading issues if
    used in combination with not thread-safe libraries.
    When reading with GDAL, a lower number of threads may be beneficial, as GDAL also performs its own threading.
    """
    default_executor_threads_jvm: int = 10

    """
    The default number of datasets that will be cached in batch jobs. A large cache will increase memory usage.
    """
    default_gdal_dataset_cache_size: int = 16

    """
    The default cache size in megabytes, which is used by GDAL. This setting applies to batch jobs.
    """
    default_gdal_cachemax: int = 150

    """
    The default maximum number of executors for batch jobs. A high number of executors can cause high costs, as allocated
    executors may be idle for a relatively high amount of time.
    We set this to 20 in kubernetes environments.
    """
    default_max_executors: int = 100

    default_driver_memory: str = "8G"

    default_driver_memoryOverhead: str = "2G"

    default_executor_memory: str = "2G"

    default_executor_memoryOverhead: str = "3G"

    default_python_memory: str = None

    default_executor_cores: int = 2

    """
    The default tile size to use for processing. By default, it is not set and the backend tries to determine a value.
    To minimize memory use, a small default size like 128 can be set. For cases with more memory per cpu, larger sizes are relevant.
    """
    default_tile_size:Optional[int] = None

    job_dependencies_poll_interval_seconds: float = 60  # poll every x seconds
    job_dependencies_max_poll_delay_seconds: float = 60 * 60 * 24 * 7  # for a maximum delay of y seconds

    udf_dependencies_pypi_index: Optional[str] = attrs.field(
        default=None, metadata={"description": "Custom PyPI index to use to install UDF dependencies"}
    )
    udf_dependencies_sleep_after_install: Optional[float] = None
    udf_dependencies_install_mode: str = attrs.field(
        # TODO: before being configurable, "direct" was original install mode, so it's the default for now.
        #       Make "disabled" the default however?
        default=UDF_DEPENDENCIES_INSTALL_MODE.DIRECT,
        validator=attrs.validators.in_(
            [
                UDF_DEPENDENCIES_INSTALL_MODE.DISABLED,
                UDF_DEPENDENCIES_INSTALL_MODE.DIRECT,
                UDF_DEPENDENCIES_INSTALL_MODE.ZIP,
            ]
        ),
    )

    """
    Only used by YARN, allows to specify paths to mount in batch job docker containers.
    """
    batch_docker_mounts: str = (
        "/var/lib/sss/pubconf/krb5.include.d:/var/lib/sss/pubconf/krb5.include.d:ro,"
        "/var/lib/sss/pipes:/var/lib/sss/pipes:rw,"
        "/usr/hdp/current/:/usr/hdp/current/:ro,"
        "/etc/hadoop/conf/:/etc/hadoop/conf/:ro,"
        "/etc/krb5.conf:/etc/krb5.conf:ro,"
        "/data/MTDA:/data/MTDA:ro,"
        "/data/projects/OpenEO:/data/projects/OpenEO:rw,"
        "/data/MEP:/data/MEP:ro,"
        "/data/users:/data/users:rw,"
        "/tmp_epod:/tmp_epod:rw,"
    )
    batch_user_docker_mounts: dict[str, List[str]] = {}
    gdalinfo_from_file: bool = True
    gdalinfo_python_call: bool = False
    gdalinfo_use_subprocess: bool = False
    gdalinfo_use_python_subprocess: bool = False

    # TODO #1009 make None by default, when appropriate configs are set.
    calrissian_config: Optional[CalrissianConfig] = CalrissianConfig()

    """
    Inject S3 profiles and tokens that allow S3 access scoped to the Job execution.
    """
    provide_s3_profiles_and_tokens: bool = False

    """
    Config file to generate identity information from this Openeo Instance
    """
    openeo_idp_details_file: Path = Path("/opt/.idp/idp_details.json")

    """
    Directory where config specific to job execution will be mounted for batch jobs.
    """
    batch_job_config_dir: Path = Path("/opt/job_config")
